### ISSUE TITLE
Current state of tests?

### ISSUE THREAD

**Issue by erniejunior** [ZeroCM/zcm#245]
I was wondering what the current state of the tests is in ZCM. 

What I found so far (after configuring with `./waf configure --use-python --use-elf --use-dev --use-clang --use-cxxtest --use-udpm -d -s` and building):

1. There is a `test.sh` file in the root dir which tells me:
```bash
This script is currently non-functional. If zcm built, it's probably working
== Running: coretest ==
./test.sh: 8: ./test.sh: ./build/test/zcm/coretest: not found
FAILED
== Running: sub-unsub-c ==
./test.sh: 8: ./test.sh: ./build/test/zcm/sub_unsub_c: not found
FAILED

...
```
So apparently it is not working. The initial note `This script is currently non-functional. If zcm built, it's probably working` is confusing to me. Does this mean that the `test.sh` script is not working at all or should it work after building?

2. I found `./build/zcm/test/runner` which tells me `Running cxxtest tests (7 tests).......OK!` when I run it. It is unclear where those CXX tests are defined?

3. There is a folder `./test` that does seem to contain some tests. But it is unclear how to build those tests or how to run them.

4. I found `./scripts/test-gen.sh` and it outputs:
```
Please provide a java package for the output java classes -- either via the "--jpkgprefix" flag or inside the type itself
ERROR: Command on line 59 failed (1): gen_zcm $f
```
5. I found [this very sad commit](https://github.com/ZeroCM/zcm/commit/9d78a7c64c8a21acca2e7309887f2f13a38649a2) apparently removing all the tests two years ago. Apparently it was removed because the tests do not compile any more.

Any guidance would be highly apreciated!

**Comment by jbendes** [comment]
This is probably better discussed over the phone haha, but I'll give it a shot here. 

There aren't many tests for zcm. This is a huge flaw in the system. We've gotten by so far with peer reviews, runtime sanitizers, and examples.

I'll split testing up into a few categories.

Core code unit tests:
Cxxtests are a fairly recent part of zcm (after 90% of the development was already done). I love the cxxtest framework and would highly recommend expanding it if you want to start writing some zcm core tests! (code found in `zcm/zcm`)

Language specific tests:
Specific to each language haha. I know of some testing frameworks for various languages, like `jest` for `javascript`.

Tooling tests:
Integration tests for `zcm-gen`, `zcm-transcoder`, etc would be nice. There are some really old ones in `test/` right now.

Transport tests:
Hard for me to even conceptualize a standard here that isn't just testing zcm itself while using various transports.

Runtime tests:
We use llvm sanitizers here to test whether or not we have threading or memory issues.


Which of these categories are you interested in writing tests for? And do you find this to be a helpful categorization?

As for the `test/` directory, the reason why this doesn't work right now is simply because we haven't figured out how to get them to compile only after `zcm-gen` has been compiled (since they rely on zcm-gen). As soon as you build, you can uncomment that line and build again, and the tests should build. Though I haven't done that in a looooong time so I wouldn't be surprised if it didn't work at all.

**Comment by erniejunior** [comment]
Those categories make sense to me. I would primarily be interested in core code tests and then transport and runtime tests. 

I will have a look into the Cxxtests. Do you also use it to test the C code? I saw that in some files you are mixing C and C++ heavily (e.g. `blocking.cpp`). How much would you object to me trying to pull this apart a bit?

I was able to get most of the tests in `./test` to run after correcting for API changes (the move from `char` to `uint8_t` for data). The coretest is a bit tricky though. I might need your support there. 
Unfortunately they all fail :-)
Are you interested in a pull request with the fixed tests? Maybe we wind a good way to make the build work. It is essentially the same issue as with the examples that require zcm to be installed before you build them.

**Comment by jbendes** [comment]
Definitely interested, though we can't enable the building of these tests by default (which is really irritating). I know waf [explicitly supports what we're trying to do with this build](https://waf.io/book/#build_compiler_first), I've just never been able to figure out how to get it to work haha. I would be all for organizing and cleaning up code, but I'm not sure to which part of the code you're specifically referring. The core of zcm (including the nonblocking framework) is c89 compatible. This is to keep zcm compatible with even the lowest level of embedded system. This is why you see a mix of c and c++ throughout the core code.

**Comment by erniejunior** [comment]
I will have a closer look at the waf book. Maybe I can make it work.

I know why you have C and C++ portions in your code and I think it is a really great idea to keep the core in C (I am using ZCM on an embedded system myself). So you can trust me that I will honor this distinction.
I was primarily referring to the blocking framework and was wondering why it is written partially in C++ since it seems to be part of the core.

**Comment by jbendes** [comment]
Honestly because it's way nicer coding in c++ than it is coding in c. And blocking transports aren't part of the c89 core. Nonblocking is the only transport included in the core (embedded-friendly) code. Blocking is posix compliant but not c89.

**Comment by olsoni** [comment]
Chiming in on a more ideal testing framework : 

We've also included some very test-like examples within the `examples` sub folder of the repo, which, as you mentioned, is only built after zcm has been fully built and installed. I think moving in that direction with the tests would be a good one because we can avoid the compile order issues. I think we should try keep some distinction between examples and tests in that examples should be built in a way that demonstrates how we would expect a user to interact with the library while the tests should be more comprehensive but not necessarily examples of best use practices.

I'll take a peak at your branch that fixes the core tests as well

**Comment by erniejunior** [comment]
I tried and failed to get WAF do build the generator first and then the tests. Unfortunately zcm-gen is far more complex (due to multiple possible output languages) than the example in the documentation. As I am not interested in diving *that* deep into WAF I decided to just move the tests in the examples folder. This way we have the usual two-step build setup with 1. build and install zcm 2. build examples and tests.

Eventually I would propose moving the examples and tests to a separate repository. There is some entanglement be means of the utils folder but apart from that there is no reason to keep the examples as a subfolder in zcm. This just gives the confusing illusion that the wscript in the root directory should also build (recurse into) the examples and tests.

What do you think?

**Comment by olsoni** [comment]
Yup, that's about where we left off with waf ... It's also worth noting that the waftools directory in examples should probably be extracted from examples since the zcmgen waftool is pretty much mandatory (for any repo that wishes to use waf & zcm that is). 

As for separating those folders out into new repos, I think I would personally prefer to keep them in the same repo, but we could definitely add a `./waf build_tests` and `./waf build_examples` command to the primary wscript to keep the build entry point the same.

**Comment by erniejunior** [comment]
Yes I already observed that the waftool folder is apparently in the wrong place. I guess I move it to the root directory?

Having a single entry point sounds great. Can you give me maybe a pointer into the waf book on how to generate those two commands? I am still learning waf ...

**Comment by olsoni** [comment]
Definitely, we're actually already doing this in zcm to set up the `./waf build_asan` and `./waf build_tsan` commands, but it's not really intuitive. See the main wscript file, the `variants` array at the top (and it's contents) are gonna be what you'll look for. Specifically, the extra commands are added here:

```
383 from waflib.Build import BuildContext, CleanContext, InstallContext, UninstallContext                
384 for x in variants:                                                                                   
385     for y in (BuildContext, CleanContext, InstallContext, UninstallContext):                         
386         name = y.__name__.replace('Context','').lower()                                              
387         class tmp(y):                                                                                
388             cmd = name + '_' + x                                                                     
389             variant = x
```

So basically, we'll want to add an "examples" and "tests" variant which recurse into their specific folders. Will likely have to trace through the main wscript to see all the places where the individual variants are treated differently, but looking for anywhere where `ctx.variant` is referenced is a good start.

**Comment by erniejunior** [comment]
thx

**Comment by erniejunior** [comment]
I did the following changes now:
* examples and tests are both in the root folder
* they are enabled using `--use-tests` and `--use-examples`
* they look for `zcm-gen` in the build folder. I am not sure if that is correct behavior because this means you first need to build zcm without examples and tests and then again with tests. Maybe this changes if we remove the zcm-gen discovery out of the configuration phase.
* they still require ZCM to be installed (and `examples/env` to be sourced) before they can even configure. Making them look for zcm-gen in the build folder was just a first step towards getting rid of the "first install then build examples" requirement. I am not done with that yet.
* I tried to adapt the travis script but failed. Unfortunately travis does not show me which step exactly failed of the script. I could not reproduce the error locally.

I was unable to use the variants mechanism to build the examples and tests because they did not allow me to influence the configuration script.

**Comment by erniejunior** [comment]
I ported most of the old tests to CXXTests. Maybe you want to have a look at it. What remains:
* I did not port the `multi_file` test because it tested some possible build errors which I was not able to reproduce. Also it was unclear how to port it to CXXTests.
* I did not port the `dispatch_loop` test because I could not understand its purpose.
* The `coretest` was too confusing with all its macros. So I left it untouched as well.

I found not solution yet for the build order issue. Nobody on the waf IRC channel answered my questions. Maybe you have a hint for me?

**Comment by erniejunior** [comment]
I fixed the build order issue by skipping all tests and examples when the `zcm-gen` is neither installed nor built. Now building the tests from scratch is an issue of just building twice. On the second run the build `zcm-gen` is detected and the tests are built as well.

I also fixed all the errors in the old build scripts that I ported to CXXTest so make them pass again and added the test runner to the travis script. 

I think we are moving towards closing the PR soon. I would now need your input regarding my above post.
